lighthouse-core/gather/computed/metrics/metric.js

```diff
@@ -0,0 +1,139 @@
+ */
+'use strict';
+
+const ComputedArtifact = require('../computed-artifact');
+const Node = require('../../../lib/dependency-graph/node');
+const Simulator = require('../../../lib/dependency-graph/simulator/simulator');
+const WebInspector = require('../../../lib/web-inspector');
+
+class MetricArtifact extends ComputedArtifact {
+  /**
+   * @param {!Node} dependencyGraph
+   * @param {function()=} condition
+   * @return {!Set<string>}
+   */
+  static getScriptUrls(dependencyGraph, condition) {
+    const scriptUrls = new Set();
+
+    dependencyGraph.traverse(node => {
+      if (node.type === Node.TYPES.CPU) return;
+      if (node.record._resourceType !== WebInspector.resourceTypes.Script) return;
+      if (condition && !condition(node)) return;
+      scriptUrls.add(node.record.url);
+    });
+
+    return scriptUrls;
+  }
+
+  /**
+   * @return {MetricCoefficients}
+   */
+  get COEFFICIENTS() {
+    throw new Error('COEFFICIENTS unimplemented!');
+  }
+
+  /**
+   * @param {!Node} dependencyGraph
+   * @param {!TraceOfTabArtifact} traceOfTab
+   * @return {!Node}
+   */
+  getOptimisticGraph(dependencyGraph, traceOfTab) { // eslint-disable-line no-unused-vars
+    throw new Error('Optimistic graph unimplemented!');
+  }
+
+  /**
+   * @param {!Node} dependencyGraph
+   * @param {!TraceOfTabArtifact} traceOfTab
+   * @return {!Node}
+   */
+  getPessimisticGraph(dependencyGraph, traceOfTab) { // eslint-disable-line no-unused-vars
+    throw new Error('Pessmistic graph unimplemented!');
+  }
+
+  /**
+   * @param {SimulationResult} simulationResult
+   * @return {number}
+   */
+  getEstimateFromSimulation(simulationResult) {
+    return simulationResult.timeInMs;
+  }
+
+  /**
+   * @param {{trace: Object, devtoolsLog: Object}} data
+   * @param {Object} artifacts
+   * @return {Promise<MetricResult>}
+   */
+  async computeMetricWithGraphs(data, artifacts, extras) {
+    const {trace, devtoolsLog} = data;
+    const graph = await artifacts.requestPageDependencyGraph({trace, devtoolsLog});
+    const traceOfTab = await artifacts.requestTraceOfTab(trace);
+    const networkAnalysis = await artifacts.requestNetworkAnalysis(devtoolsLog);
+
+    const optimisticGraph = this.getOptimisticGraph(graph, traceOfTab, extras);
+    const pessimisticGraph = this.getPessimisticGraph(graph, traceOfTab, extras);
+
+    // TODO(phulce): use rtt and throughput from config.settings instead of defaults
+    const options = {
+      additionalRttByOrigin: networkAnalysis.additionalRttByOrigin,
+      serverResponseTimeByOrigin: networkAnalysis.serverResponseTimeByOrigin,
+    };
+
+    const optimisticEstimate = new Simulator(optimisticGraph, options).simulate();
+    const pessimisticEstimate = new Simulator(pessimisticGraph, options).simulate();
+
+    optimisticEstimate.timeInMs = this.getEstimateFromSimulation(
```

Comment:
we're redefining the `timeInMs`? Feels a little weird.  What does the `timeInMs` from .simulate represent?

Reply:
yeah, from .simulate it's a shortcut for `Math.max(...nodeTiming.map(timing => timing.endTime))`

we want to return our optimistic/pessimistic estimates in ms along with the node timing that led to that computation, open to other suggestions on how to structure the return value